content/blog/example-of-mapreduce-with-ezmeral-data-fabric-mapr-database-binary-table.md

```diff
@@ -0,0 +1,365 @@
+
+
+
+**Important note**:
+
+<ins>the commands in this article are all executed as the user "mapr", which is by default the admin user of the EDF Development Environment.</ins>
+
+<ins>You can also use the "root" user to create the table and run the application, but if you don't modify the ACEs of the table, the "mapr" user would be not able to see the data in the table.</ins>
+
+
+
+We are going to use the [hbase shell][HbaseShell] to create a [Binary Table][BinaryTable] inside the EDF Database.
+To be able to use the `hbase shell`, we need to install the **[mapr-hbase][MaprHbasePackage]** package first.
+
+
+
+#### For convenience of data management, I would like to create a volume for the Binary Table, the command are as following:
+
+
+
+```bash
+sudo -u mapr maprcli volume list -output terse -columns volumename,volumetype,actualreplication,localpath,mounted,mountdir,logicalUsed,used,nameContainerDataThresholdMB,nameContainerSizeMB,needsGfsck
+
+maprcli volume create -name test.binarytable1 -path /testbinarytable1volume
+
+# sudo -u mapr maprcli volume mount -name test.binarytable1 -path /testbinarytable1volume
+sudo -u mapr hadoop fs -ls -d -h /testbinarytable1volume
+sudo -u mapr hadoop mfs -ls /testbinarytable1volume
+```
+
+
+
+☝ The volume's name is: test.binarytable1 and it will be mounted as <ins>/testbinarytable1volume/</ins> in the EDF File System.
+
+
+
+#### Now we can create the Binary Table
+
+
+
+👇 Inside `hbase shell`:
+
+
+
+```
+create '/testbinarytable1volume/notifications','attributes','metrics'
+```
+
+
+
+Table name is <ins>/testbinarytable1volume/notifications</ins>. **attributes** and **metrics** are column families.
+
+
+
+**Note**: In EDF Database Binary Table, the table name is by default a path in the File System.
+You can change the style of the table name to be like in Apache HBase, refer to👉: [Mapping to HBase Table Namespaces][MappingToHbaseTableNamespaces].
+
+
+
+### Build and Run The MapReduce Application
+
+
+
+**Important note**: Before running the MapReduce application, if you are using the Development Environment, you have to complete the following steps: [Installing Hadoop and YARN][InstallingHadoopAndYarn].
+
+
+It's because by default the YARN framework is not installed and hence the MapReduce framework is not installed.
+
+
+
+#### MapReduce Application Source Code
+
+
+
+I have put the code on Github: [Example-MapReduce-With-EzmeralDataFabricMapR-DatabaseBinaryTable][MapReduceEDFDataBaseBinaryTable].
+You can download it and compile it using Visual Studio Code.
+
+
+
+This MapReduce application is simple, the following is the logic:
+
+
+
+1. Get data from the source table: <ins>/testbinarytable1volume/notifications</ins> and output aggregated data to the target table: <ins>/testbinarytable1volume/summary</ins>.
+2. It's basically a varietas of "Word Count". This MapReduce application simply aggregates the number of rows which contains a column called "type" in the "attributes" column family.
+
+
+

```

Comment:
For a better rendering of the next 2 sentences, I would suggest to move them left.